File: Sources/AISecurity/Core/SenderWhitelist.swift
import Foundation

/// Manages sender whitelist — trusted senders skip low-confidence checks
/// but ALWAYS get scanned for malicious attachments, URLs, malware, and prompt injection.
///
/// Industry standard (Gmail, Outlook, Proofpoint): whitelisting reduces noise
/// from social engineering/urgency alerts but never bypasses malware scanning.
final class SenderWhitelist: @unchecked Sendable {

    // MARK: - Types

    struct Entry: Codable, Sendable {
        let address: String           // exact address or @domain
        let source: Source
        let addedAt: String           // ISO8601
        var note: String?

        enum Source: String, Codable, Sendable {
            case userExplicit = "user"
            case contactsSync = "contacts"
        }
    }

    /// Scan policy for a sender based on whitelist status
    struct ScanPolicy: Sendable {
        let isWhitelisted: Bool

        /// Categories that should ALWAYS fire regardless of whitelist.
        /// `authority_impersonation` lives here because BEC / CEO-fraud attacks
        /// specifically target mailboxes inside trusted domains — if anything,
        /// a trusted sender asking for a wire transfer is MORE suspicious, not
        /// less. We never silence these alerts.
        static let alwaysScanCategories: Set<String> = [
            "malicious_url", "dangerous_attachment", "malware_dropper",
            "prompt_injection", "crypto_scam",
            "authority_impersonation"
        ]

        /// Categories suppressed for whitelisted senders. Limited to generic
        /// tone/urgency signals (social_engineering) which are noisy on real
        /// business mail from known contacts. Never suppress anything that
        /// represents actionable financial or credential risk.
        static let suppressedCategories: Set<String> = [
            "social_engineering"
        ]

        /// Categories with raised threshold for whitelisted senders
        static let reducedCategories: Set<String> = [
            "phishing", "sensitive_data_request"
        ]

        /// Should this threat category fire for this sender?
        func shouldAlert(category: String, intentLayers: Int) -> Bool {
            // Always-scan categories fire regardless
            if Self.alwaysScanCategories.contains(category) { return true }

            // Unknown senders — everything fires
            if !isWhitelisted { return true }

            // Whitelisted: suppress generic social-engineering noise only.
            if Self.suppressedCategories.contains(category) { return false }

            // Whitelisted: raise threshold for reduced categories (need 5+ layers)
            if Self.reducedCategories.contains(category) {
                return intentLayers >= 5
            }

            // All other categories: fire normally
            return true
        }
    }

    // MARK: - Properties

    private var entries: [String: Entry] = [:]   // lowercased address → entry
    private let filePath: String
    private let lock = NSLock()

    /// Freemail domains that cannot be whitelisted at domain level
    private static let freemailDomains: Set<String> = [
        "gmail.com", "googlemail.com", "outlook.com", "hotmail.com", "live.com",
        "yahoo.com", "ymail.com", "icloud.com", "me.com", "mac.com", "aol.com",
        "protonmail.com", "proton.me", "zoho.com", "mail.com", "gmx.com", "yandex.com"
    ]

    /// Path to encrypted whitelist
    private let encryptedFilePath: String
    /// Path to legacy plain JSON whitelist (for migration)
    private let legacyFilePath: String

    // MARK: - Init

    init(securityDir: String) {
        self.filePath = (securityDir as NSString).appendingPathComponent("whitelist.enc")
        self.encryptedFilePath = self.filePath
        self.legacyFilePath = (securityDir as NSString).appendingPathComponent("whitelist.json")
        load()
    }

    // MARK: - Public API

    /// Check if a sender address is whitelisted
    func policy(for sender: String) -> ScanPolicy {
        let addr = extractAddress(sender).lowercased()
        lock.lock()
        defer { lock.unlock() }

        // Check exact address match
        if entries[addr] != nil {
            return ScanPolicy(isWhitelisted: true)
        }

        // Check domain match
        if let atIdx = addr.firstIndex(of: "@") {
            let domain = String(addr[addr.index(after: atIdx)...])
            if entries["@" + domain] != nil {
                return ScanPolicy(isWhitelisted: true)
            }
        }

        return ScanPolicy(isWhitelisted: false)
    }

    /// Add a sender to the whitelist
    @discardableResult
    func add(_ sender: String, source: Entry.Source = .userExplicit, note: String? = nil) -> Bool {
        let addr = extractAddress(sender).lowercased()
        guard !addr.isEmpty else { return false }

        // Block domain-level whitelist for freemail providers
        if addr.hasPrefix("@") {
            let domain = String(addr.dropFirst())
            if Self.freemailDomains.contains(domain) { return false }
        }

        let entry = Entry(
            address: addr,
            source: source,
            addedAt: ISO8601DateFormatter().string(from: Date()),
            note: note
        )

        lock.lock()
        entries[addr] = entry
        lock.unlock()
        save()
        return true
    }

    /// Remove a sender from the whitelist
    func remove(_ sender: String) {
        let addr = extractAddress(sender).lowercased()
        lock.lock()
        entries.removeValue(forKey: addr)
        lock.unlock()
        save()
    }

    /// Get all whitelist entries
    func allEntries() -> [Entry] {
        lock.lock()
        defer { lock.unlock() }
        return Array(entries.values).sorted { $0.addedAt > $1.addedAt }
    }

    /// Check if a sender is whitelisted
    func isWhitelisted(_ sender: String) -> Bool {
        policy(for: sender).isWhitelisted
    }

    // MARK: - Encrypted Persistence

    private func load() {
        // Try encrypted file first (whitelist.enc), using the current master key.
        if FileManager.default.fileExists(atPath: encryptedFilePath),
           let hexData = FileManager.default.contents(atPath: encryptedFilePath),
           let hex = String(data: hexData, encoding: .utf8)?.trimmingCharacters(in: .whitespacesAndNewlines),
           !hex.isEmpty {

            // Path A: decrypt with the current master key.
            if let json = SecurityCoreBridge.decryptWhitelist(hex),
               let jsonData = json.data(using: .utf8),
               let arr = try? JSONDecoder().decode([Entry].self, from: jsonData) {
                lock.lock()
                for entry in arr {
                    entries[entry.address.lowercased()] = entry
                }
                lock.unlock()
                return
            }

            // Path B: master key failed — try the legacy default passphrase
            // (data written by pre-master-key versions of the app). If this
            // succeeds, re-encrypt immediately with the master key so the
            // legacy-decryption path is never used again on this install.
            if let json = SecurityCoreBridge.decryptWhitelistLegacy(hex),
               let jsonData = json.data(using: .utf8),
               let arr = try? JSONDecoder().decode([Entry].self, from: jsonData) {
                lock.lock()
                for entry in arr {
                    entries[entry.address.lowercased()] = entry
                }
                lock.unlock()
                save()   // re-encrypt under the master key
                return
            }

            // Both paths failed — whitelist is unreadable. Do NOT silently
            // clobber it; leave the file in place so an operator can inspect.
            // Start with an empty in-memory whitelist this run.
            return
        }

        // Fall back to legacy plain JSON (whitelist.json) for first-ever migration
        if FileManager.default.fileExists(atPath: legacyFilePath),
           let data = FileManager.default.contents(atPath: legacyFilePath),
           let arr = try? JSONDecoder().decode([Entry].self, from: data) {
            lock.lock()
            for entry in arr {
                entries[entry.address.lowercased()] = entry
            }
            lock.unlock()
            // Migrate: save as encrypted, then remove legacy file
            save()
            try? FileManager.default.removeItem(atPath: legacyFilePath)
            return
        }
    }

    private func save() {
        lock.lock()
        let arr = Array(entries.values)
        lock.unlock()

        // Encrypt and save. Fail CLOSED — never write plaintext to disk.
        // If encryption fails (master key missing / crypto core broken), log
        // to the diagnostic file and leave the existing encrypted file alone
        // rather than silently downgrading security.
        guard let jsonData = try? JSONEncoder().encode(arr),
              let json = String(data: jsonData, encoding: .utf8),
              let hex = SecurityCoreBridge.encryptWhitelist(json) else {
            let diagPath = (NSHomeDirectory() as NSString).appendingPathComponent(".mac-security/logs/whitelist-errors.log")
            let msg = "[\(Date())] ERROR: whitelist encryption failed — changes NOT persisted. Master key likely unset.\n"
            if let handle = FileHandle(forWritingAtPath: diagPath) {
                handle.seekToEndOfFile()
                handle.write(msg.data(using: .utf8) ?? Data())
                handle.closeFile()
            } else {
                try? msg.write(toFile: diagPath, atomically: true, encoding: .utf8)
            }
            return
        }

        try? hex.data(using: .utf8)?.write(to: URL(fileURLWithPath: encryptedFilePath))
    }

    // MARK: - Helpers

    /// Extract bare email address from "Name <email@example.com>" format
    private func extractAddress(_ sender: String) -> String {
        if let start = sender.firstIndex(of: "<"),
           let end = sender.firstIndex(of: ">") {
            return String(sender[sender.index(after: start)..<end])
        }
        return sender.trimmingCharacters(in: .whitespaces)
    }
}

File: SecurityCore/crates/security-core/src/sender_whitelist.rs
use serde::{Deserialize, Serialize};
use std::collections::HashMap;

/// Source of the whitelist entry.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub enum EntrySource {
    #[serde(rename = "user")]
    UserExplicit,
    #[serde(rename = "contacts")]
    ContactsSync,
}

/// A whitelist entry — exact address or @domain.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct WhitelistEntry {
    pub address: String,
    pub source: EntrySource,
    #[serde(rename = "addedAt")]
    pub added_at: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub note: Option<String>,
}

/// Scan policy for a sender based on whitelist status.
#[derive(Debug, Clone)]
pub struct ScanPolicy {
    pub is_whitelisted: bool,
}

/// Categories that should ALWAYS fire regardless of whitelist.
pub const ALWAYS_SCAN_CATEGORIES: &[&str] = &[
    "malicious_url",
    "dangerous_attachment",
    "malware_dropper",
    "prompt_injection",
    "crypto_scam",
];

/// Categories suppressed for whitelisted senders.
pub const SUPPRESSED_CATEGORIES: &[&str] = &[
    "social_engineering",
    "authority_impersonation",
];

/// Categories with raised threshold for whitelisted senders.
pub const REDUCED_CATEGORIES: &[&str] = &[
    "phishing",
    "sensitive_data_request",
];

impl ScanPolicy {
    /// Should this threat category fire for this sender?
    pub fn should_alert(&self, category: &str, intent_layers: u8) -> bool {
        // Always-scan categories fire regardless
        if ALWAYS_SCAN_CATEGORIES.contains(&category) {
            return true;
        }

        // Unknown senders — everything fires
        if !self.is_whitelisted {
            return true;
        }

        // Whitelisted: suppress social engineering / authority impersonation
        if SUPPRESSED_CATEGORIES.contains(&category) {
            return false;
        }

        // Whitelisted: raise threshold for reduced categories (need 5+ layers)
        if REDUCED_CATEGORIES.contains(&category) {
            return intent_layers >= 5;
        }

        // All other categories: fire normally
        true
    }
}

/// Freemail domains that cannot be whitelisted at domain level.
pub const FREEMAIL_DOMAINS: &[&str] = &[
    "gmail.com", "googlemail.com", "outlook.com", "hotmail.com", "live.com",
    "yahoo.com", "ymail.com", "icloud.com", "me.com", "mac.com", "aol.com",
    "protonmail.com", "proton.me", "zoho.com", "mail.com", "gmx.com", "yandex.com",
];

/// In-memory sender whitelist with JSON persistence.
pub struct SenderWhitelist {
    entries: HashMap<String, WhitelistEntry>,
    file_path: String,
}

impl SenderWhitelist {
    /// Create a new whitelist, loading from disk if the file exists.
    pub fn new(security_dir: &str) -> Self {
        let file_path = format!("{}/whitelist.json", security_dir);
        let mut wl = SenderWhitelist {
            entries: HashMap::new(),
            file_path,
        };
        wl.load();
        wl
    }

    /// Create from a specific file path.
    pub fn from_path(file_path: &str) -> Self {
        let mut wl = SenderWhitelist {
            entries: HashMap::new(),
            file_path: file_path.to_string(),
        };
        wl.load();
        wl
    }

    /// Check the scan policy for a sender.
    pub fn policy(&self, sender: &str) -> ScanPolicy {
        let addr = extract_address(sender).to_lowercase();

        // Check exact address match
        if self.entries.contains_key(&addr) {
            return ScanPolicy { is_whitelisted: true };
        }

        // Check domain match
        if let Some(at_idx) = addr.find('@') {
            let domain_key = format!("@{}", &addr[at_idx + 1..]);
            if self.entries.contains_key(&domain_key) {
                return ScanPolicy { is_whitelisted: true };
            }
        }

        ScanPolicy { is_whitelisted: false }
    }

    /// Add a sender to the whitelist. Returns false if blocked (empty or freemail domain).
    pub fn add(
        &mut self,
        sender: &str,
        source: EntrySource,
        note: Option<&str>,
    ) -> bool {
        let addr = extract_address(sender).to_lowercase();
        if addr.is_empty() {
            return false;
        }

        // Block domain-level whitelist for freemail providers
        if let Some(domain) = addr.strip_prefix('@') {
            if FREEMAIL_DOMAINS.contains(&domain) {
                return false;
            }
        }

        let entry = WhitelistEntry {
            address: addr.clone(),
            source,
            added_at: chrono::Utc::now().to_rfc3339(),
            note: note.map(|s| s.to_string()),
        };

        self.entries.insert(addr, entry);
        self.save();
        true
    }

    /// Remove a sender from the whitelist.
    pub fn remove(&mut self, sender: &str) {
        let addr = extract_address(sender).to_lowercase();
        self.entries.remove(&addr);
        self.save();
    }

    /// Get all whitelist entries sorted by added_at descending.
    pub fn all_entries(&self) -> Vec<&WhitelistEntry> {
        let mut entries: Vec<_> = self.entries.values().collect();
        entries.sort_by(|a, b| b.added_at.cmp(&a.added_at));
        entries
    }

    /// Check if a sender is whitelisted.
    pub fn is_whitelisted(&self, sender: &str) -> bool {
        self.policy(sender).is_whitelisted
    }

    // -- Persistence --

    fn load(&mut self) {
        let data = match std::fs::read_to_string(&self.file_path) {
            Ok(d) => d,
            Err(_) => return,
        };
        let arr: Vec<WhitelistEntry> = match serde_json::from_str(&data) {
            Ok(a) => a,
            Err(_) => return,
        };
        for entry in arr {
            self.entries.insert(entry.address.to_lowercase(), entry);
        }
    }

    fn save(&self) {
        let arr: Vec<&WhitelistEntry> = self.entries.values().collect();
        if let Ok(data) = serde_json::to_string_pretty(&arr) {
            let _ = std::fs::write(&self.file_path, data);
        }
    }
}

/// Extract bare email address from "Name <email@example.com>" format.
fn extract_address(sender: &str) -> String {
    if let Some(start) = sender.find('<') {
        if let Some(end) = sender.find('>') {
            if start < end {
                return sender[start + 1..end].to_string();
            }
        }
    }
    sender.trim().to_string()
}

#[cfg(test)]
mod tests {
    use super::*;

    fn make_whitelist() -> SenderWhitelist {
        // Use a non-existent path so no file is loaded
        SenderWhitelist {
            entries: HashMap::new(),
            file_path: "/tmp/test_whitelist_nonexistent.json".to_string(),
        }
    }

    #[test]
    fn unknown_sender_not_whitelisted() {
        let wl = make_whitelist();
        let policy = wl.policy("unknown@evil.com");
        assert!(!policy.is_whitelisted);
    }

    #[test]
    fn add_and_check_exact() {
        let mut wl = make_whitelist();
        assert!(wl.add("friend@example.com", EntrySource::UserExplicit, None));
        assert!(wl.is_whitelisted("friend@example.com"));
        assert!(!wl.is_whitelisted("stranger@example.com"));
    }

    #[test]
    fn add_and_check_domain() {
        let mut wl = make_whitelist();
        assert!(wl.add("@company.com", EntrySource::UserExplicit, Some("Work")));
        assert!(wl.is_whitelisted("anyone@company.com"));
        assert!(!wl.is_whitelisted("other@evil.com"));
    }

    #[test]
    fn block_freemail_domain() {
        let mut wl = make_whitelist();
        assert!(!wl.add("@gmail.com", EntrySource::UserExplicit, None));
        assert!(!wl.is_whitelisted("anyone@gmail.com"));
    }

    #[test]
    fn allow_freemail_exact_address() {
        let mut wl = make_whitelist();
        assert!(wl.add("friend@gmail.com", EntrySource::UserExplicit, None));
        assert!(wl.is_whitelisted("friend@gmail.com"));
        assert!(!wl.is_whitelisted("other@gmail.com"));
    }

    #[test]
    fn extract_from_angle_brackets() {
        let addr = extract_address("John Doe <john@example.com>");
        assert_eq!(addr, "john@example.com");
    }

    #[test]
    fn extract_bare_address() {
        let addr = extract_address("john@example.com");
        assert_eq!(addr, "john@example.com");
    }

    #[test]
    fn remove_entry() {
        let mut wl = make_whitelist();
        wl.add("test@example.com", EntrySource::UserExplicit, None);
        assert!(wl.is_whitelisted("test@example.com"));
        wl.remove("test@example.com");
        assert!(!wl.is_whitelisted("test@example.com"));
    }

    #[test]
    fn policy_always_scan_fires() {
        let policy = ScanPolicy { is_whitelisted: true };
        assert!(policy.should_alert("malicious_url", 0));
        assert!(policy.should_alert("dangerous_attachment", 0));
        assert!(policy.should_alert("crypto_scam", 0));
    }

    #[test]
    fn policy_suppresses_social_engineering() {
        let policy = ScanPolicy { is_whitelisted: true };
        assert!(!policy.should_alert("social_engineering", 3));
        assert!(!policy.should_alert("authority_impersonation", 3));
    }

    #[test]
    fn policy_reduced_needs_five_layers() {
        let policy = ScanPolicy { is_whitelisted: true };
        assert!(!policy.should_alert("phishing", 4));
        assert!(policy.should_alert("phishing", 5));
        assert!(!policy.should_alert("sensitive_data_request", 3));
        assert!(policy.should_alert("sensitive_data_request", 5));
    }

    #[test]
    fn policy_unknown_sender_fires_all() {
        let policy = ScanPolicy { is_whitelisted: false };
        assert!(policy.should_alert("social_engineering", 0));
        assert!(policy.should_alert("phishing", 1));
        assert!(policy.should_alert("authority_impersonation", 0));
    }

    #[test]
    fn case_insensitive() {
        let mut wl = make_whitelist();
        wl.add("Test@Example.COM", EntrySource::UserExplicit, None);
        assert!(wl.is_whitelisted("test@example.com"));
        assert!(wl.is_whitelisted("TEST@EXAMPLE.COM"));
    }
}
